### PR TITLE
Add support for cast storage on same stypes

### LINKED ISSUE #8524: Cast storage to same stype gives an error for csr
Note: Providing complete information in the most concise form is the best way to get help. This issue template serves as the checklist for essential information to most of the technical issues.

If the issue is non-technical, feel free to present the information in what you believe is the best form.

## Description
Cast storage to same stype gives an error for csr

## Environment info (Required)
AWS EC2 Instance based on DLAMI

```
mx.nd.sparse.csr_matrix([[0,1,0],[1,2,1],[3,4,5]]).tostype('csr')
```

Package used (Python/R/Scala/Julia):
Python

MXNet commit hash:
23f5af34398e29236d3d0489c137832e1696fae9

## Error Message:
Stack trace returned 10 entries:
[bt] (0) /home/ubuntu/sparse_support/mxnet/python/mxnet/../../lib/libmxnet.so(_ZN4dmlc15LogMessageFatalD1Ev+0x3f) [0x7f67a635789f]
[bt] (1) /home/ubuntu/sparse_support/mxnet/python/mxnet/../../lib/libmxnet.so(_ZN5mxnet2op22CastStorageComputeImplIN7mshadow3cpuEEEvRKNS_9OpContextERKNS_7NDArrayES9_+0x25b) [0x7f67a74adc73]
[bt] (2) /home/ubuntu/sparse_support/mxnet/python/mxnet/../../lib/libmxnet.so(_ZN5mxnet2op20CastStorageComputeExIN7mshadow3cpuEEEvRKN4nnvm9NodeAttrsERKNS_9OpContextERKSt6vectorINS_7NDArrayESaISC_EERKSB_INS_9OpReqTypeESaISH_EESG_+0x30e) [0x7f67a74a95e0]
[bt] (3) /home/ubuntu/sparse_support/mxnet/python/mxnet/../../lib/libmxnet.so(_ZNSt17_Function_handlerIFvRKN4nnvm9NodeAttrsERKN5mxnet9OpContextERKSt6vectorINS4_7NDArrayESaIS9_EERKS8_INS4_9OpReqTypeESaISE_EESD_EPSJ_E9_M_invokeERKSt9_Any_dataS3_S7_SD_SI_SD_+0x91) [0x7f67a65d150b]
[bt] (4) /home/ubuntu/sparse_support/mxnet/python/mxnet/../../lib/libmxnet.so(_ZNKSt8functionIFvRKN4nnvm9NodeAttrsERKN5mxnet9OpContextERKSt6vectorINS4_7NDArrayESaIS9_EERKS8_INS4_9OpReqTypeESaISE_EESD_EEclES3_S7_SD_SI_SD_+0xa6) [0x7f67a7fb08a2]
[bt] (5) /home/ubuntu/sparse_support/mxnet/python/mxnet/../../lib/libmxnet.so(_ZZN5mxnet10imperative14PushFComputeExERKSt8functionIFvRKN4nnvm9NodeAttrsERKNS_9OpContextERKSt6vectorINS_7NDArrayESaISA_EERKS9_INS_9OpReqTypeESaISF_EESE_EEPKNS2_2OpES5_RKNS_7ContextERKS9_IPNS_6engine3VarESaISW_EES10_RKS9_INS_8ResourceESaIS11_EERKS9_IPSA_SaIS16_EES1A_SJ_ENKUlNS_10RunContextENSU_18CallbackOnCompleteEE_clES1B_S1C_+0xca) [0x7f67a7faada6]
[bt] (6) /home/ubuntu/sparse_support/mxnet/python/mxnet/../../lib/libmxnet.so(_ZNSt17_Function_handlerIFvN5mxnet10RunContextENS0_6engine18CallbackOnCompleteEEZNS0_10imperative14PushFComputeExERKSt8functionIFvRKN4nnvm9NodeAttrsERKNS0_9OpContextERKSt6vectorINS0_7NDArrayESaISF_EERKSE_INS0_9OpReqTypeESaISK_EESJ_EEPKNS7_2OpESA_RKNS0_7ContextERKSE_IPNS2_3VarESaIS10_EES14_RKSE_INS0_8ResourceESaIS15_EERKSE_IPSF_SaIS1A_EES1E_SO_EUlS1_S3_E_E9_M_invokeERKSt9_Any_dataOS1_OS3_+0x67) [0x7f67a7fb57e6]
[bt] (7) /home/ubuntu/sparse_support/mxnet/python/mxnet/../../lib/libmxnet.so(_ZNKSt8functionIFvN5mxnet10RunContextENS0_6engine18CallbackOnCompleteEEEclES1_S3_+0x67) [0x7f67a85ec57d]
[bt] (8) /home/ubuntu/sparse_support/mxnet/python/mxnet/../../lib/libmxnet.so(_ZN5mxnet6engine14ThreadedEngine15ExecuteOprBlockENS_10RunContextEPNS0_8OprBlockE+0x2e7) [0x7f67a85ee4b9]
[bt] (9) /home/ubuntu/sparse_support/mxnet/python/mxnet/../../lib/libmxnet.so(_ZN5mxnet6engine23ThreadedEnginePerDevice9CPUWorkerILN4dmlc19ConcurrentQueueTypeE0EEEvNS_7ContextEPNS1_17ThreadWorkerBlockIXT_EEE+0x72) [0x7f67a85fe6e0]

[01:06:54] /home/ubuntu/sparse_support/mxnet/dmlc-core/include/dmlc/./logging.h:308: [01:06:54] src/engine/./threaded_engine.h:370: [01:06:54] src/operator/tensor/./cast_storage-inl.h:346: Not implemented

## Minimum reproducible example
```
import mxnet as mx
mx.nd.sparse.csr_matrix([[0,1,0],[1,2,1],[3,4,5]]).tostype('csr')
```

## What have you tried to solve it?
1. https://github.com/apache/incubator-mxnet/blob/master/src/operator/tensor/cast_storage-inl.h#L333 The cases source and dest as same stype is not handled here.

### PR DESCRIPTION
**Pull request by anirudh2290**
## Description ##
Fixes #8524

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: src/operator/tensor/cast_storage-inl.h [hunk 1]
```diff
@@ -328,6 +329,53 @@ void CastStorageCsrDnsImpl(const OpContext& ctx,
...
+void CastStorageRspRspImpl(const OpContext& ctx,
+                           const NDArray& rsp,
+                           NDArray* output) {
+  CHECK_EQ(rsp.storage_type(), output->storage_type())
+      << "Copying with different storage type";
+  mshadow::Stream<xpu>* s = ctx.get_stream<xpu>();
+  if (!rsp.storage_initialized()) {
+    FillZerosRspImpl(s, *output);
+    return;
+    }
```
Comment (eric-haibin-lin) [inline src/operator/tensor/cast_storage-inl.h]: indentation?

File: src/operator/tensor/cast_storage-inl.h [hunk 2]
```diff
@@ -328,6 +329,53 @@ void CastStorageCsrDnsImpl(const OpContext& ctx,
...
+template<typename xpu>
+void CastStorageCsrCsrImpl(const OpContext& ctx,
+                           const NDArray& csr,
+                           NDArray* output) {
+  mshadow::Stream<xpu>* s = ctx.get_stream<xpu>();
+  if (!csr.storage_initialized()) {
+    FillZerosCsrImpl(s, *output);
+    return;
+  }
+  output->CheckAndAllocAuxData(csr::kIndPtr, csr.aux_shape(csr::kIndPtr));
```
Comment (eric-haibin-lin) [inline src/operator/tensor/cast_storage-inl.h]: nit: `CheckAndAlloc(const std::vector<TShape> &aux_shapes)` will make it neater..

**Comment by eric-haibin-lin** [comment]
Do you mind also updating the doc for the added support? https://mxnet.incubator.apache.org/versions/master/api/python/ndarray/sparse.html#mxnet.ndarray.sparse.cast_storage ?

**Comment by anirudh2290** [comment]
@eric-haibin-lin Addressed.

**Comment by haojin2** [comment]
LGTM!

**Pull request merged**